dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/DataQueryServiceTest.java

```diff
@@ -0,0 +1,141 @@
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hisp.dhis.test.TestBase.createOrganisationUnit;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.Set;
+import org.hisp.dhis.analytics.AnalyticsSecurityManager;
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.analytics.DataQueryService;
+import org.hisp.dhis.common.BaseIdentifiableObject;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.common.UserOrgUnitType;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.user.User;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class DataQueryServiceTest {
+  private static OrganisationUnit ouA;
+
+  private static OrganisationUnit ouB;
+
+  private static OrganisationUnit ouC;
+
+  private static OrganisationUnit ouD;
+
+  private static DataQueryParams dataQueryParams;
+
+  @BeforeAll
+  static void setUp() {
+    ouA = createOrganisationUnit('A');
+    ouB = createOrganisationUnit('B');
+    ouC = createOrganisationUnit('C');
+    ouD = createOrganisationUnit('D');
+    dataQueryParams =
+        DataQueryParams.newBuilder().withUserOrgUnitType(UserOrgUnitType.DATA_OUTPUT).build();
+  }
+
+  @Test
+  void testGetUserOrgUnitsWithExplicitGrantedForAnalyticsOrganisationUnits() {
+    // given
+    User currentUser = mock(User.class);
+    AnalyticsSecurityManager analyticsSecurityManager = mock(AnalyticsSecurityManager.class);
+    when(currentUser.getDataViewOrganisationUnits()).thenReturn(Set.of(ouB, ouC, ouD));
+    when(analyticsSecurityManager.getCurrentUser(dataQueryParams)).thenReturn(currentUser);
+    DataQueryService dataQueryService =
+        new DefaultDataQueryService(
+            mock(DimensionalObjectProducer.class),
+            mock(IdentifiableObjectManager.class),
+            analyticsSecurityManager);
+
+    // when
+    List<OrganisationUnit> userOrgUnits = dataQueryService.getUserOrgUnits(dataQueryParams, null);
+
+    // then
+    assertEquals(3, userOrgUnits.size());
+    assertThat(
+        userOrgUnits.stream().map(BaseIdentifiableObject::getName).toList(),
+        containsInAnyOrder("OrganisationUnitB", "OrganisationUnitC", "OrganisationUnitD"));
+  }
+
+  @Test
+  void testGetUserOrgUnitsWithDeniedForAnalyticsOrganisationUnits() {
+    // given
+    User currentUser = mock(User.class);
+    AnalyticsSecurityManager analyticsSecurityManager = mock(AnalyticsSecurityManager.class);
+    when(currentUser.getDataViewOrganisationUnits()).thenReturn(Set.of());
+    when(currentUser.getDataViewOrganisationUnits()).thenReturn(Set.of(ouA));
```

Comment:
Something is not right here. It seems duplicated.

I'll ping you for other comments.

Reply:
done